### ISSUE TITLE
Support RETURNING in moor files

### ISSUE THREAD

**Issue by simolus3** [simolus3/moor#1096]
sqlite3 version 3.35.0, which was released yesterday, added support for `RETURNING` clauses. `RETURNING` clauses can be added to an `UPDATE`, `DELETE` or `INSERT` statement and allow evaluating expressions for affected rows:

```sql
CREATE TABLE t0(
  a INTEGER PRIMARY KEY,
  b DATE DEFAULT CURRENT_TIMESTAMP,
  c INTEGER
);

newT: INSERT INTO t0(c) VALUES(random()) RETURNING *;
```

Here, `newT` would return a single row return the rowid, the the current timestamp and the generated random value.

This is an awesome feature, let's support it in moor! Adding support for this requires a considerable effort across the analyzer, the generator and database implementations:

- [ ] Support this in `sqlparser`
- [ ] Support this in `moor_generator`
- [ ] Support this in sqflite-based implementations
- [ ] Support this in `moor/ffi.dart`
- [ ] Support this in the web implementation

Another interesting approach would be to add a version of `into(table).insert` that automatically returns the full data class by generating a `RETURNING` clause. It probably shouldn't be the default since it requires a super new sqlite3 version, but we could have an extension or a new method for that.

**Comment by fgblomqvist** [comment]
Was there ever a version of `into(table).insert` added that will return the full object that was added?

**Comment by simolus3** [comment]
Yes! You can use `into(table).`[`insertReturning`](https://pub.dev/documentation/moor/latest/moor/InsertStatement/insertReturning.html).

**Comment by fgblomqvist** [comment]
Ah sweet!

**Comment by kuhnroyal** [comment]
@simolus3 Is there a way to get this working on Desktop with `VmDatabase.memory()`?
I started using this but now it is failing in some test cases.

**Comment by simolus3** [comment]
@kuhnroyal Is the error about `RETURNING` in general? sqlite 3.35 is very recent, so the environment where you run your tests on may not have it yet.
On Linux (and probably on macOS too) you could download sqlite3 somewhere and then run the tests with `LD_LIBRARY_PATH=/folder/containing/your/sqlite flutter test`.

**Comment by kuhnroyal** [comment]
Yea the error is about `RETURNING` being unknown it seems. I am on latest MacOS 11.3.1. Will try a custom library path tomorrow. Thanks!

**Comment by kuhnroyal** [comment]
That doesn't seem to work.
The lib here does not exist on my system:
```dart
  if (Platform.isMacOS) {
    return DynamicLibrary.open('/usr/lib/libsqlite3.dylib');
  }
```

It seems to be using the one that is installed with the Android SDK in `sdk/platform-tools/sqlite3`, I probably have that in my path.

I tried to download the binary from here: https://www.sqlite.org/download.html but the binaries for Mac don't seem to contain a lib.

I was hoping I could somehow handle this with a download and `open.overrideFor(OperatingSystem.macOS, () => DynamicLibrary.open(library.absolute.path))`.

**Comment by simolus3** [comment]
You can compile sqlite3 manually on macOS (I tried this on an arm mac, but I think the output format is universal):

```
mkdir /tmp/sqlite
cd /tmp/sqlite
curl https://sqlite.org/2021/sqlite-amalgamation-3350500.zip --output sqlite.zip
unzip sqlite.zip
cd sqlite-amalgamation-3350500
cc -dynamiclib -o libsqlite3.dylib sqlite3.c -arch x86_64 -arch arm64 -O3
```

You can then copy the generated `libsqlite3.dylib` somewhere and use an opening override in your test code. It looks like `LD_PATH` does not work as expected.

**Comment by kuhnroyal** [comment]
This works, thanks, not sure how I use this in CI yet :)

**Comment by simolus3** [comment]
I guess you can upload your compiled `libsqlite3.dylib`  somewhere, fetch that in the CI and override the open behavior accordingly. Not great, but it should serve as a temporary workaround until macOS upgrades sqlite3.

**Comment by simolus3** [comment]
Can you reproduce that with a simpler example (or with table definitions that you could share)? Problem 1 sounds like our type inference got lost, problem 2 sounds more bizarre to me. It's hard to say without seeing the tables, but it may also be something weird with sqlite3.

**Comment by svarlet** [comment]
That's my moor file stripped off of everything irrelevant:
```sql
CREATE TABLE sync_sessions (
   id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
   endpoint TEXT NOT NULL
);

CREATE TABLE downloads (
    id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
    uri TEXT NOT NULL MAPPED BY `const UriConverter()`,
    last_modified_time INT MAPPED BY `const InstantConverter()`,
    attempts_started_count INT NOT NULL DEFAULT 0,
    priority INT NOT NULL DEFAULT 0,
    last_attempt_timestamp INT,
    status TEXT NOT NULL,
    handed_off bool NOT NULL DEFAULT false,
    sync_session INT NOT NULL,

    UNIQUE (sync_session, uri)
);

CREATE TABLE successes (
    id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
    download INT NOT NULL UNIQUE
);

CREATE TABLE failures (
    id INT NOT NULL PRIMARY KEY AUTOINCREMENT,
    download INT NOT NULL UNIQUE
);

checkoutNextDownloadables:
    UPDATE downloads
    SET status = 'CHECKED_OUT'
    FROM (
        SELECT d.id
        FROM downloads AS d
        INNER JOIN sync_sessions AS ss ON d.sync_session = ss.id
        LEFT JOIN successes AS s ON s.download = d.id
        LEFT JOIN failures AS f ON f.download = d.id
        WHERE d.handed_off = 0 AND
            s.download IS NULL AND
            f.download IS NULL AND
            ( (d.status = 'TEMPORARY_FAILURE' AND 
                d.attempts_started_count < 3 AND 
                d.last_attempt_timestamp >= date('now', '-10 second') ) OR 
                d.status IN ('NOT_STARTED', 'INCOMPLETE') ) 
        ORDER BY d.priority ASC,
                d.last_attempt_timestamp ASC
        LIMIT :count
        ) as w
    WHERE downloads.id = w.id
    RETURNING downloads.id;
```

**Comment by simolus3** [comment]
Thanks a lot for the helpful repo! This query returns a list of `null`s on sqlite3 version 3.35.4 and 3.35.5, but it returns the expected value on 3.36.0. So it looks like you stumbled upon a now-fixed sqlite3 bug. I'll push the latest sqlite3 to `sqlite3_flutter_libs` soon.

Finally, I discovered two bugs in how we resolve `RETURNING` clauses in the `sqlparser` package and fixed them with c83b630514770a4621b2230ac12ea10a9b333fe0. This will be part of the next moor release (for which I unfortunately don't have an ETA yet).

**Comment by svarlet** [comment]
Great, thanks for investigating!
That query was an attempt to replace some code that performed 1 read + 19 writes (in a batch) by 1 write and 19 selects. This is because I'm getting a lot of `SqliteException: database is locked` at runtime and my guess is that we are attempting too many concurrent writes. Though if you can't provide an ETA at this time, I suppose I'll have to try something else.

**Comment by Vedsaga** [comment]
> [`insertReturning`](https://pub.dev/documentation/moor/latest/moor/InsertStatement/insertReturning.html).

this link return 404
![image](https://user-images.githubusercontent.com/59083528/164968864-7613e199-9d11-4903-a41d-0aae25aa86ea.png)

**Comment by Vedsaga** [comment]
I guess the correct link would be https://drift.simonbinder.eu/api/drift/insertstatement/insertreturning
